src/Workleap.DomainEventPropagation.Subscription.PullDelivery.Tests/EventPullerTests.cs

```diff
@@ -0,0 +1,270 @@
+﻿using AutoBogus;
+using Azure.Messaging;
+using FakeItEasy;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.Extensions.Options;
+using Workleap.DomainEventPropagation.EventGridClientAdapter;
+
+namespace Workleap.DomainEventPropagation.Subscription.PullDelivery.Tests;
+
+// Test class cannot be made static
+#pragma warning disable CA1052
+public class EventPullerTests
+#pragma warning restore CA1052
+{
+    private static IEventGridClientAdapter GivenFakeClient(IEventGridClientWrapperFactory clientWrapperFactory, string subName)
+    {
+        var fakeClient = A.Fake<IEventGridClientAdapter>();
+        A.CallTo(() => clientWrapperFactory.CreateClient(subName)).Returns(fakeClient);
+        return fakeClient;
+    }
+
+    private static IServiceScopeFactory GivenScopeFactory(ICloudEventHandler handler)
+    {
+        var scopeFactory = A.Fake<IServiceScopeFactory>();
+        var scope = A.Fake<IServiceScope>();
+        var serviceCollection = new ServiceCollection();
+        serviceCollection.AddScoped<ICloudEventHandler>(_ => handler);
+        A.CallTo(() => scopeFactory.CreateScope()).Returns(scope);
+        A.CallTo(() => scope.ServiceProvider).Returns(serviceCollection.BuildServiceProvider());
+        return scopeFactory;
+    }
+
+    private static EventPropagationSubscriptionOptions GivenEventPropagationSubscriptionOptions(IOptionsMonitor<EventPropagationSubscriptionOptions> optionMonitor, string clientName)
+    {
+        var option = AutoFaker.Generate<EventPropagationSubscriptionOptions>();
+        A.CallTo(() => optionMonitor.Get(clientName)).Returns(option);
+        return option;
+    }
+
+    private static void RegisterCloudEventsInClient(IEventGridClientAdapter client, EventPropagationSubscriptionOptions option, params EventGridClientAdapter.EventGridClientAdapter.EventBundle[] events)
+    {
+        A.CallTo(() => client.ReceiveCloudEventsAsync(
+            option.TopicName,
+            option.SubscriptionName,
+            A<CancellationToken>._)).Returns(events);
+    }
+
+    private static async Task StartWaitAndStop(IHostedService puller)
+    {
+// We need this to start on the thread pool otherwise it will just block the test
+#pragma warning disable CS4014
+        Task.Run(() => puller.StartAsync(CancellationToken.None));
+        await Task.Delay(50);
+#pragma warning restore CS4014
+        await puller.StopAsync(CancellationToken.None);
+    }
+
+    public class TwoSubscribers : EventPullerTests
+    {
+        [Fact]
+        public async Task GivenPuller_WhenStarted_ThenEveryRegisteredClientWasCalled()
+        {
+            // Given
+            var clientName1 = AutoFaker.Generate<string>();
```

Comment:
C'est quoi l'avantage entre ça et écrire une string?

Reply:
Using [`AutoBogus`](https://github.com/nickdodd79/AutoBogus) will generate a non-empty string with the highlighted call.

Generally speaking, we should not care about what test `string`s like this one hold but we do care that it has a non-empty value. We could also hardcode a value, but the content shouldn't really matter.

As a side note, we could probably define `private` constants for that test class given most tests rely on those client names being defined.

Reply:
As discussed, I will keep this _as-is_ for the time being.